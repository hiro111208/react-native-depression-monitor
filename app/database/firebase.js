//import firebase from "firebase/app";
import "firebase/firestore";
import firebase from 'firebase'
require('firebase/auth')

const firebaseConfig = {
  apiKey: "AIzaSyDwOUkdxC7kNNILoFTSx05PTUxRci_kcaw",
  authDomain: "fireship-tutorial-646fc.firebaseapp.com",
  projectId: "fireship-tutorial-646fc",
  storageBucket: "fireship-tutorial-646fc.appspot.com",
  messagingSenderId: "289064122528",
  appId: "1:289064122528:web:ee39abfa2183c8499071c5",
  measurementId: "G-D35RDH9V0E",
};

firebase.initializeApp(firebaseConfig);

export default firebase;
