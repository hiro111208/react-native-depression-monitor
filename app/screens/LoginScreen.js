import React, { useState } from "react";
import {
  StyleSheet,
  Text,
  View,
  TextInput,
  Alert,
  ActivityIndicator,
  TouchableOpacity,
} from "react-native";
import firebase from "../database/firebase";
import colors from "../config/colors";

/*
  Screen where users can login to access their dashboards
*/
function LoginScreen(props) {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [isLoading, setIsLoading] = useState(false);
  const [errorMessage, setErrorMessage] = useState("");
  const [isVerified, setIsVerified] = useState(true);

  const db = firebase.firestore();

  /**
   * Locates the progress of the user based on their uid
   */
  function navigateUser(uid) {
    var docRef = db.collection("users").doc(uid);
    docRef
      .get()
      .then((doc) => {
        if (doc.exists) {
          const userProgress = doc.data();

          // A category hasn't been dropped
          if (userProgress.categoryDropped == "NONE") {
            props.navigation.navigate("DemoScreen", {
              user: userProgress,
            });
            setIsLoading(false);
          }

          // Proceeds straight to the dashboard (skipping category and demo)
          else {
            props.navigation.navigate("PatientDashboard");
            setIsLoading(false);
          }
        }

        // doc.data() will be undefined in this case
        else {
          console.log("No such document!");
        }
      })
      .catch((error) => {
        console.log("Error getting document:", error);
      });
  }

  //Send verification email to user to allow them to login
  const sendVerificationEmail = () => {
    firebase
      .auth()
      .currentUser.sendEmailVerification()
      .then(function () {
        console.log("email verification sent");
        setErrorMessage(
          `Please verify your email through the link we've sent to: ${email}`
        );
      })
      .catch(function (error) {
        //console.log("failed to send email verification");
        console.log(error.code);
        setIsLoading(false);
        setErrorMessage(error.message);
      });
  };

  //Verify user credentials with firebase. If user has verified their email
  //send user to dashboard if not show button to allow user to send a new verification email
  //if email used is admin email redirect to admin dashboard
  const userLogin = () => {
    if (email === "" && password === "") {
      Alert.alert("Enter details to login!");
    } else {
      setIsLoading(true);
      firebase
        .auth()
        .signInWithEmailAndPassword(email, password)
        .then((res) => {
          if (firebase.auth().currentUser.email == "admin@joyapp.com") {
            reset();
            props.navigation.navigate("AdminDashboard");
            setIsLoading(false);
          } else {
              if (firebase.auth().currentUser.emailVerified) {
                console.log("User logged-in successfully!");
                reset();
                navigateUser(firebase.auth().currentUser.uid);
              } else {
                  setIsLoading(false);
                  setErrorMessage("Your email has not been verified");
                  setIsVerified(false);
              }
            }
        })
        .catch((error) => {
          console.log(error.code);
          setIsLoading(false);

          switch(error.code) {
            case "auth/invalid-email":
              setErrorMessage(`Please enter a valid email.`);
              break;
            case "auth/wrong-password":
              setErrorMessage(`Incorrect password for '${email}'.`);
              break;
            case "auth/user-not-found":
              setErrorMessage(`No account for '${email}' exists, please sign up with this email.`);
              break;
            default:
              setErrorMessage(error.message);
              break;
          }
        });
    }
  };

  //reset all states
  const reset = () => {
    setEmail("");
    setPassword("");
    setIsVerified(true);
    setErrorMessage("");
  };

  //While loading show preloader
  if (isLoading) {
    return (
      <View style={styles.preloader}>
        <ActivityIndicator size="large" color="#9E9E9E" />
      </View>
    );
  }
  //Render the login screen interface
  return (
    <View style={styles.container}>
      {/*Render input fields for email and password*/}
      <TextInput
        style={styles.inputStyle}
        placeholder="Email"
        value={email}
        keyboardType="email-address"
        onChangeText={(userEmail) => setEmail(userEmail)}
        testID={"TEST_ID_EMAIL_INPUT"}
      />
      <TextInput
        style={styles.inputStyle}
        placeholder="Password"
        value={password}
        onChangeText={(userPassword) => setPassword(userPassword)}
        maxLength={15}
        secureTextEntry={true}
        testID={"TEST_ID_PASSWORD_INPUT"}
      />   

      {/*Render text that shows error  messages */}
      <Text testID={"TEST_ID_MESSAGE"} style={{color:"red"}}>{errorMessage}</Text>

      {/*Render login button which calls userLogin method and checks credentials in input fields*/}
      <TouchableOpacity
        activeOpacity={0.5}
        style={styles.loginButton}
        onPress={()=>userLogin()}
        testID={"TEST_ID_LOGIN_BUTTON"}
      >
        <Text style= {styles.signInText}>
          LOG IN
        </Text>
      </TouchableOpacity>

      {/*Render text to allow user to go to sign up screen*/}
      <Text
        style={styles.textButton}
        onPress={() => {
          reset(); 
          props.navigation.navigate("SignupScreen");
        }}
        testID={"TEST_ID_SIGNUP_BUTTON"}
        >
        Don't have an account? Click here to signup
      </Text>

      {/*Render text to allow user to go to forgot password screen*/}
      <Text
        style={styles.textButton}
        onPress={() => {
          reset();
          props.navigation.navigate("ForgotPasswordScreen");
        }}
        testID={"TEST_ID_FORGOT_BUTTON"}
        >
        Forgot your password?
      </Text>

      {/*Render button to send new verification email */}
      <View> 
      {!isVerified
        ? <TouchableOpacity
            activeOpacity = { 0.5 }
            style={styles.loginButton}
            onPress={()=>sendVerificationEmail()}
            testID={"TEST_ID_VERIFY_BUTTON"}
          >
          <Text style= {styles.signInText}>
            Send verification email
          </Text>
        </TouchableOpacity>
        :
        null}
      </View>
    </View>
  );
}

export default LoginScreen;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    padding: 35,
    backgroundColor: "#fff",
  },
  inputStyle: {
    width: "100%",
    marginBottom: 15,
    marginTop: 15,
    paddingBottom: 20,
    alignSelf: "center",
    borderColor: "#ccc",
    borderBottomWidth: 1.5,
  },
  loginButton: {
    width: 300,
    backgroundColor: colors.darkBorder,
    alignSelf: "center",
    marginTop: 10,
    paddingTop: 15,
    paddingBottom: 15,
    borderRadius: 50,
  },
  textButton: {
    color: colors.darkBorder,
    marginTop: 25,
    textAlign: "center",
  },
  preloader: {
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    position: "absolute",
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#fff",
  },
  signInText: {
    color: "white",
    textAlign: "center",
    fontSize: 15,
  },
});
