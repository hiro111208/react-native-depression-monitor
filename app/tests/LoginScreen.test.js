import React from "react";
import renderer from "react-test-renderer";
import LoginScreen from "../screens/LoginScreen";
import { Alert } from "react-native";
import { render, fireEvent, waitFor } from "@testing-library/react-native";

import firebase from "../../firebase.js"
//import * as firebaseEmulator from "../../node_modules/@firebase/testing"; //unused firebase emulator import

describe("Testing LoginScreen.js", () => {

    beforeAll(async () => {
        jest.setTimeout(10000); 
        await firebase.auth().createUserWithEmailAndPassword("login@test.com", "password");
    });

    afterAll(async () => {
        await firebase.auth().currentUser.delete().catch(error => {
            credential = firebase.auth.EmailAuthProvider.credential("login@test.com","password");
            firebase.auth().currentUser.reauthenticateWithCredential(credential);
            firebase.auth().currentUser.delete();
        });
    })

    test("LoginScreen renders correctly", () => {
        const tree = renderer.create(<LoginScreen />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("Expect to show enter details alert with no credentials", async () => {
        const { getByTestId } = render(<LoginScreen />);
        const button = getByTestId("TEST_ID_LOGIN_BUTTON");

        jest.spyOn(Alert, "alert");
        await waitFor(() => {
            fireEvent.press(button);
            expect(Alert.alert).toHaveBeenCalledWith("Enter details to login!");
        })
    });

    test("signInWithEmailAndPassword should be unsuccessful with incorrect email format", async () => {
        let error = "";
        try {
            await firebase.auth().signInWithEmailAndPassword("notAProperEmail", "password");
            // console.log("Test 2 passed");
        } catch (err) {
            error = err.toString();
            // console.log("Catch error 2: " + error);
        }
        expect(error).toEqual("Error: The email address is badly formatted.");
    });

    test("signInWithEmailAndPassword should be successful with correct and verified credentials", async () => {
        let error = "";
        try {
            const newUser = await firebase.auth().signInWithEmailAndPassword("login@test.com", "password");
            expect(newUser.user).toBeTruthy();
            // console.log(newUser.user.emailVerified);
            // console.log("Test 3 passed");
        } catch(err) {
            // console.log("Catch error 3: " + error);
            error = err.toString();
        }
        expect(error).toEqual("");
    });

    test("signInWithEmailAndPassword should be unsuccessful with no credentials", async () => {
        let error = "";
        try {
            await firebase.auth().signInWithEmailAndPassword("", "");
        } catch (err) {
            error = err.toString();
        }
        expect(error).toEqual("Error: The email address is badly formatted.");
    }); 

    test("signInWithEmailAndPassword should be unsuccessful with incorrect password", async () => {
        let error = "";
        try {
            await firebase.auth().signInWithEmailAndPassword("login@test.com", "incorrect");
        } catch (err) {
            error = err.toString();
        }
        expect(error).toEqual("Error: The password is invalid or the user does not have a password.");
    });

    test("signInWithEmailAndPassword with incorrect password should display message", async () => {
        const { getByTestId } = render(<LoginScreen />)
        const emailInput = getByTestId("TEST_ID_EMAIL_INPUT");
        const passwordInput = getByTestId("TEST_ID_PASSWORD_INPUT");
        const button = getByTestId("TEST_ID_LOGIN_BUTTON");

        jest.spyOn(firebase.auth(), "signInWithEmailAndPassword");
        await fireEvent.changeText(emailInput, "login@test.com");
        await fireEvent.changeText(passwordInput, "wrongpassword");           
        await fireEvent.press(button);  
        
        await waitFor(() => {
            expect(firebase.auth().signInWithEmailAndPassword).toHaveBeenCalled();
            expect("The password is invalid or the user does not have a password.").toEqual(getByTestId("TEST_ID_MESSAGE").props.children) 
        });
    });

    test("signInWithEmailAndPassword should be unsuccessful with unverified email", async () => {
        const { getByTestId } = render(<LoginScreen />)
        const emailInput = getByTestId("TEST_ID_EMAIL_INPUT");
        const passwordInput = getByTestId("TEST_ID_PASSWORD_INPUT");
        const button = getByTestId("TEST_ID_LOGIN_BUTTON");

        jest.spyOn(firebase.auth(), "signInWithEmailAndPassword");
        await fireEvent.changeText(emailInput, "login@test.com");
        await fireEvent.changeText(passwordInput, "password");           
        await fireEvent.press(button);  
        
        await waitFor(() => {
            expect(firebase.auth().signInWithEmailAndPassword).toHaveBeenCalled();
            expect("Your email has not been verfied").toEqual(getByTestId("TEST_ID_MESSAGE").props.children) 
        });
    });
});

//currently unused firebase emulator set up and tear down
// beforeAll(async () => {
    //Firebase emulator set-up, not currently in use
    // firebaseEmulator.initializeTestApp({
    //     projectId: "fireship-tutorial-646fc",
    //     auth: { uid: "login-test", email: "login@test.com", password:"password" } 
    //});
// });

// afterAll( async () => {
    //Promise.all(firebaseEmulator.apps().map(app => app.delete()))
// })
