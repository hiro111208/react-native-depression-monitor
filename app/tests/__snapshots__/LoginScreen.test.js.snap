// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing LoginScreen.js LoginScreen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "display": "flex",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
      "padding": 35,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Email"
    rejectResponderTermination={true}
    style={
      Object {
        "alignSelf": "center",
        "borderBottomWidth": 1,
        "borderColor": "#ccc",
        "marginBottom": 15,
        "paddingBottom": 15,
        "width": "100%",
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    maxLength={15}
    onChangeText={[Function]}
    placeholder="Password"
    rejectResponderTermination={true}
    secureTextEntry={true}
    style={
      Object {
        "alignSelf": "center",
        "borderBottomWidth": 1,
        "borderColor": "#ccc",
        "marginBottom": 15,
        "paddingBottom": 15,
        "width": "100%",
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Text
    style={
      Object {
        "color": "red",
      }
    }
  >
    
  </Text>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#ffa351ff",
        "borderRadius": 50,
        "marginTop": 10,
        "opacity": 1,
        "paddingBottom": 15,
        "paddingTop": 15,
        "width": 300,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 15,
          "textAlign": "center",
        }
      }
    >
      SIGN IN
    </Text>
  </View>
  <Text
    onPress={[Function]}
    style={
      Object {
        "color": "#ffa351ff",
        "marginTop": 25,
        "textAlign": "center",
      }
    }
  >
    Don't have an account? Click here to signup
  </Text>
  <Text
    onPress={[Function]}
    style={
      Object {
        "color": "#ffa351ff",
        "marginTop": 25,
        "textAlign": "center",
      }
    }
  >
    Forgot your password?
  </Text>
  <View />
</View>
`;
